ldap-connector/src/com/axonivy/connector/ldap/util/JndiConfig.java

```diff
@@ -0,0 +1,285 @@
+      env.put(Context.SECURITY_PRINCIPAL, userName);
+      env.put(Context.SECURITY_CREDENTIALS, password);
+    } else {
+      env.put(Context.SECURITY_AUTHENTICATION, AUTH_KIND_NONE);
+    }
+
+    if (useSsl) {
+      env.put(Context.SECURITY_PROTOCOL, "ssl");
+    }
+    if (useLdapConnectionPool) {
+      env.put(LDAP_CONNECTION_POOL, "true");
+    }
+
+    env.put(Context.REFERRAL, "follow");
+
+    if (JndiProvider.ACTIVE_DIRECTORY.equals(this.provider)) {
+      // fix for active directory bug.
+      // See more details in class ch.ivyteam.naming.ldap.ldapURLContextFactory
+      env.put(Context.URL_PKG_PREFIXES, "ch.ivyteam.naming");
+    }
+    return env;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    JndiConfig jndiConfig;
+    if (obj instanceof JndiConfig) {
+      jndiConfig = (JndiConfig) obj;
+
+      if (this == jndiConfig) {
+        return true;
+      }
+
+      return ((url.equals(jndiConfig.url)) &&
+              (provider.equals(jndiConfig.provider)) &&
+              (authenticationKind.equals(jndiConfig.authenticationKind)) &&
+              (defaultContext.equals(jndiConfig.defaultContext)) &&
+              (userName.equals(jndiConfig.userName)) &&
+              (password.equals(jndiConfig.password)) &&
+              (useSsl == jndiConfig.useSsl));
+    }
+    return false;
+  }
+
+  @Override
+  public String toString() {
+    return "JndiConfig[url=" + url + ", "
+            + "provider=" + provider + ", "
+            + "authenticationKind=" + authenticationKind + ","
+            + "useSSL=" + useSsl + ","
+            + "defaultContext=" + defaultContext + "]";
+  }
+
+  @Override
+  public int hashCode() {
+    return getUrl().hashCode();
+  }
+
+  public boolean isUseLdapConnectionPool() {
+    return useLdapConnectionPool;
+  }
+
+  public void setUseLdapConnectionPool(boolean _useLdapConnectionPool) {
+    this.useLdapConnectionPool = _useLdapConnectionPool;
+  }
+
+  public static Builder create() {
+    return new Builder();
+  }
+
+  public static Builder create(JndiConfig config) {
+    return new Builder()
+            .authenticationKind(config.authenticationKind)
+            .defaultContext(config.defaultContext)
+            .password(config.password)
+            .provider(config.provider)
+            .url(config.url)
+            .useLdapConnectionPool(config.useLdapConnectionPool)
+            .userName(config.userName)
+            .useSsl(config.useSsl);
+  }
+
+  public static final class Builder {
+    private JndiProvider provider = new JndiProvider(null, null);
+    private String url = "";
+    private String authenticationKind = "none";
+    private String userName = "";
+    private String password = "";
+    private String defaultContext = "";
+    private boolean useSsl = false;
+    private boolean useLdapConnectionPool = false;
+
+    public Builder provider(@SuppressWarnings("hiding") JndiProvider provider) {
```

Comment:
I usually add @SuppressWarnings("hiding") on the builder class itself.
so you don't need to add to every method.